### PR TITLE
Flow the cancellation tokens to ReadAsync and WriteAsync

### PR DESCRIPTION
**Pull request by davidfowl**

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http/MessageBody.cs [hunk 1]
```diff
@@ -43,7 +43,7 @@ public virtual async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationT
 
             while (true)
```
Comment (gfoidl) [inline src/Kestrel.Core/Internal/Http/MessageBody.cs]: Should cancellation checked here too?

I.e.
```c#
while (!cancellationToken.IsCancellationRequested)
{
    // ....
}
```

Same below on line 84.

Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http/MessageBody.cs]: Eh, then I'd have to throw in if `cancellationToken.ThrowIfCancellationRequested()`. I'd also need to handle the case where the loop never ran and the token got triggered.

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/MessageBody.cs]: Makes sense -- I missed the throwing-behavior. 

In
https://github.com/aspnet/KestrelHttpServer/blob/105c395bac37186b789496a5656848bf318eb8db/src/Kestrel.Core/Internal/Http/MessageBody.cs#L104
the token could be passed in too. (Can't comment direct on that line here).

Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http/MessageBody.cs]: good catch

**Comment by Tratcher** [comment]
Just because or did you run into a specific issue?

**Comment by davidfowl** [comment]
I was trying things and it didn't work as expected then I realized that we don't pass in these tokens anywhere.

**Comment by Tratcher** [comment]
Sounds like you need some new tests.

**Comment by davidfowl** [comment]
Yea, I'll add some.

File: src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs [hunk 2]
```diff
@@ -89,7 +89,7 @@ public void Write<T>(Func<PipeWriter, T, long> callback, T state)
             }
         }
 
-        public Task WriteAsync<T>(Func<PipeWriter, T, long> callback, T state)
+        public Task WriteAsync<T>(Func<PipeWriter, T, long> callback, T state, CancellationToken cancellationToken)
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs]: This is only used for headers and therefore isn't used.

File: src/Kestrel.Core/Internal/Http/IHttpOutputProducer.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
     public interface IHttpOutputProducer : IDisposable
     {
         void Abort(ConnectionAbortedException abortReason);
-        Task WriteAsync<T>(Func<PipeWriter, T, long> callback, T state);
+        Task WriteAsync<T>(Func<PipeWriter, T, long> callback, T state, CancellationToken cancellationToken);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/IHttpOutputProducer.cs]: The caller never has a cancellationToken.

**Review by halter73 — CHANGES_REQUESTED** [review]
We need at least 1 regression test.

File: src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs (hunk 2 above)
Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs]: Why do you say that? Theoretically, you can change the response buffer to a small number and cause WriteAsync to hang on FlushAsync here https://github.com/aspnet/KestrelHttpServer/blob/80c8e83c689a8e2ed52e1c4b7b84c4c2d8ad15ee/src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs#L107

File: src/Kestrel.Core/Internal/Http/IHttpOutputProducer.cs (hunk 3 above)
Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http/IHttpOutputProducer.cs]: Can you elaborate? WriteChunkedAsync calls this method today.

File: src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs (hunk 2 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Http/Http1OutputProducer.cs]: My bad, for some reason I thought this was (only) called by WriteResponseHeaders which would never use a CancellationToken. Forgot this was used to write chunked response bodies.

Looking at the code now, I realize we should also just remove the `Write<T>` overload. That's what we used to use for header writing prior to the HTTP/2 refactoring.

It looks like we regressed CancellationToken handling for chunked response bodies in 2.1.0. Maybe we should backport this fix.

File: src/Kestrel.Core/Internal/Http/IHttpOutputProducer.cs (hunk 3 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Http/IHttpOutputProducer.cs]: See above response. I was thinking about the old code.

**Comment by davidfowl** [comment]
Bleh, flaky tests and leaks a result of the exception. I'll get back to this...

**Review by Tratcher — APPROVED** [review]

File: test/Kestrel.InMemory.FunctionalTests/RequestTests.cs [hunk 4]
```diff
@@ -55,6 +56,73 @@ public async Task StreamsAreNotPersistedAcrossRequests()
...
+                    context.Response.ContentLength = 14;
+                    await context.Response.WriteAsync("Read cancelled");
+                }
+
+            }, new TestServiceContext(LoggerFactory)))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.Send(
+                        "GET / HTTP/1.1",
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply test/Kestrel.InMemory.FunctionalTests/RequestTests.cs]: Hah, good catch.

**Comment by halter73** [comment]
Can you remove the sync `Http1OutputProducer.Write<T>(Func<PipeWriter, T, long> callback, T state)` method as part of this change?

**Review by halter73 — APPROVED** [review]

**Comment by davidfowl** [comment]
@dotnet-bot test Windows Debug x64 Build

File: test/Kestrel.InMemory.FunctionalTests/ResponseTests.cs [hunk 5]
```diff
@@ -110,12 +110,12 @@ public async Task ResponseBodyWriteAsyncCanBeCancelled()
...
             using (var server = new TestServer(async context =>
             {
-                await context.Response.WriteAsync("hello", cts.Token);
+                await context.Response.WriteAsync("hello", cts.Token).DefaultTimeout();
                 try
                 {
                     var task = context.Response.WriteAsync("world", cts.Token);
                     Assert.False(task.IsCompleted);
-                    await task;
+                    await task.DefaultTimeout();
```
Comment (halter73) [inline test/Kestrel.InMemory.FunctionalTests/ResponseTests.cs]: 👍

**Comment by davidfowl** [comment]
@dotnet-bot test OSX 10.12 Debug Build

**Comment by davidfowl** [comment]
@dotnet-bot test OSX 10.12 Release Build

**Pull request merged**